Add API reference (#29)
orochi-network/orochi-network.github.io: src/orochi-network/api-reference.md

# API Reference

## Initial

> We assumed that you have an initialized instance of Orand

Reviewer: ```suggestion
We assume that you have an initialized instance of Orand.
```

```ts
import { orand } from "@orochi-network/sdk";

const orandInstance = new orand.Orand({
  url: "https://orand-test-service.orochi.network/",
  user: "YOUR_REGISTERED_USERNAME",
  secretKey: "YOUR_REGISTERED_HMAC_SECRET",
  consumerAddress: "YOUR_APPLICATION_SMART_CONTRACT_ADDRESS",
  chaiId: 1,
});
```

**Note:**

> To prevent replay attack, we're required you to input `consumerAddress`. It's smart contract address that will consume the randomness. Each `consumerAddress` will have different epoch target, you only able to to submit the epoch which has the matched epoch with target epoch on `OrandProviderV1` smart contract.

Author: ```suggestion
To prevent replay attack, we're required you to input `consumerAddress`. It's a smart contract address that will consume the randomness. Each `consumerAddress` will have different epoch target, you only able to to submit the epoch which has the matched epoch with target epoch on `OrandProviderV1` smart contract.
```

> To prevent replay attack, we're required you to input `consumerAddress`. It's smart contract address that will consume the randomness. Each `consumerAddress` will have different epoch target, you only able to to submit the epoch which has the matched epoch with target epoch on `OrandProviderV1` smart contract.

Reviewer: ```suggestion
To prevent replay attack, we're required you to input `consumerAddress`. It's a smart contract address that will consume the randomness. Each `consumerAddress` will have different epoch target, you are only able to submit the epoch which has the matched epoch with target epoch on `OrandProviderV1` smart contract.
```

> Some results will be return in `BigInt` due to new update from `ethers.js`.

Reviewer: ```suggestion
Some results will be returned in `BigInt` due to new update from `ethers.js`.
```

### .getPublicEpoch()

```ts
await orandInstance.getPublicEpoch(0);
```

> Allowed you to get a list of public epoch at the starting `epoch`, it will return an array of `IOrandEpoch`. The result will be limited to 20 records to prevent Denial of Service.

Reviewer: ```suggestion
Allowed you to get a list of public epochs at the starting `epoch`, it will return an array of `IOrandEpoch`. The result will be limited to 20 records to prevent Denial of Service.
```

> Public epoch has `receiverAddress` is `0x0000000000000000000000000000000000000000`, submit public epoch to `OrandProviderV1` won't have any affect since the consumer address is `0x0000000000000000000000000000000000000000`.

Reviewer: ```suggestion
Public epoch has `receiverAddress` is `0x0000000000000000000000000000000000000000`, submitting public epoch to `OrandProviderV1` won't have any affect since the consumer address is `0x0000000000000000000000000000000000000000`.
```

  gamma: '8a3059cec8687c2d9d7048098a8484b0ebb8839d6589be070affac5a7763dd8e73a316231302534ea834e897835f610d61bc4dd8a2b75b71be35414b2fb2a2ea',
  c: '9345c77ac9c7c1ba6d084ccf6997e2fcf623dc9cdc72e42a91855b1c8cc1c5fa',
  s: '8457bc202e735627d5a27133744bf3bb0dd72fb5e85ce826d0716c107fd44430',
  y: '147c78180c05d041a9e8b3bb11cf59f1c871019db940532bd38df22f8194f28c',
  witnessAddress: '6dfb0007085713f2621380670e8578eb83df552b',
  witnessGamma: 'a191f86cff2b88e06b38c64e4d2ddfb3c0673b32c7e229bd46f6a2262a962f700765aba72c8d225133874d0f26327a6851a67946b2deb1d7c722d4836d4cd1f4',
  witnessHash: '0142eeaf5d41dad6a3b2f42f0a06ae8b17b47084c78bcd86f44693a264a9defb72a9b3ee2a3466810bb69e6360ca02f16b32f5f241ae7089651bc8f6a1f580c4',
  inverseZ: '77024d6c015a9b0cbe758246843d7d8712115abaac53f761a6e1594d240a8a02',
  signatureProof: '6f7e2c4eef8a3bbc0699971ab2d80693702356f1873d45b20829b50873943e8975f6f978d1daafc10c2c7ecda8bf3dda823485cfa7d17740398852423188f54c1b000000000000000000000003f16f07cfd6e9ac06925fcf68dd0b450f4131989d147c78180c05d041a9e8b3bb11cf59f1c871019db940532bd38df22f8194f28c',
  createdDate: '2023-03-10 06:56:08'
}
```

### .getPrivateEpoch()

```ts
await orandInstance.getPrivateEpoch(3);
```

> Allowed you to get a list of private epoch the starting `epoch`, it will return an array of `IOrandEpoch`. The result will be limited to 20 records to prevent Denial of Service. These result will be bound to `receiverAddress`.

Reviewer: ```suggestion
Allowed you to get a list of private epochs at the starting `epoch`, it will return an array of `IOrandEpoch`. The result will be limited to 20 records to prevent Denial of Service. These result will be bounded to `receiverAddress`.
```

    c: '9345c77ac9c7c1ba6d084ccf6997e2fcf623dc9cdc72e42a91855b1c8cc1c5fa',
    s: '8457bc202e735627d5a27133744bf3bb0dd72fb5e85ce826d0716c107fd44430',
    y: '147c78180c05d041a9e8b3bb11cf59f1c871019db940532bd38df22f8194f28c',
    witnessAddress: '6dfb0007085713f2621380670e8578eb83df552b',
    witnessGamma: 'a191f86cff2b88e06b38c64e4d2ddfb3c0673b32c7e229bd46f6a2262a962f700765aba72c8d225133874d0f26327a6851a67946b2deb1d7c722d4836d4cd1f4',
    witnessHash: '0142eeaf5d41dad6a3b2f42f0a06ae8b17b47084c78bcd86f44693a264a9defb72a9b3ee2a3466810bb69e6360ca02f16b32f5f241ae7089651bc8f6a1f580c4',
    inverseZ: '77024d6c015a9b0cbe758246843d7d8712115abaac53f761a6e1594d240a8a02',
    signatureProof: '6f7e2c4eef8a3bbc0699971ab2d80693702356f1873d45b20829b50873943e8975f6f978d1daafc10c2c7ecda8bf3dda823485cfa7d17740398852423188f54c1b000000000000000000000003f16f07cfd6e9ac06925fcf68dd0b450f4131989d147c78180c05d041a9e8b3bb11cf59f1c871019db940532bd38df22f8194f28c',
    createdDate: '2023-03-10 06:56:08'
  }
]
```

### .verifyECDSAProof()

```ts
await orandInstance.verifyECDSAProof(epochData);
```

> Allowed you verify the ECDSA proof of any given epoch.

Reviewer: ```suggestion
Allowed you to verify the ECDSA proof of any given epoch.
```

**Result:**

```txt
{
  signer: '0x7e9e03a453867a7046B0277f6cD72E1B59f67a0e',
  receiverEpoch: 2n,
  receiverAddress: '0xF16F07cfd6e9Ac06925FCf68dD0b450f4131989D',
  y: 91097723859136686723473270573409338368251757405505259439091937259103551628289n
}
```

- `signer` MUST be equal to `await orandInstance.getOperatorAddress()` to make sure the ECDSA proof is valid.
- `receiverEpoch` the required epoch number that need to be submit in the next publication.
- `receiverAddress` MUST be equal to predefine `receiverAddress` in the initial stage
- `y` Result of the current epoch

### .getActiveChain()

```ts
await orandInstance.getActiveChain();
```

> Allowed you read the data of current active chain.

Reviewer: ```suggestion
Allowed you to read the data of current active chain.
```

**Result:**

```txt
{
  url: 'https://data-seed-prebsc-1-s2.binance.org:8545',
  providerAddress: '0x75C0e60Ca5771dd58627ac8c215661d0261D5D76'
}
```

- `url` is the JSON-PRC provider, if you don't want you can use another JSON-RPC provider
- `providerAddress` is `OrandProviderV1` address on active chain

### .getOperatorAddress()

```ts
await orandInstance.getOperatorAddress();
```

> Get the operator's address that going to sign the ECDSA Proof.

Reviewer: ```suggestion
Get the operator's address that is going to sign the ECDSA Proof.
```

**Result:**

```txt
0x7e9e03a453867a7046B0277f6cD72E1B59f67a0e
```

### .getReceiverAlpha()

```ts
await orandInstance.getReceiverAlpha(
  "0xf16f07cfd6e9ac06925fcf68dd0b450f4131989d"
);
```

Get current `alpha` of receiver, current `alpha` is previous the result of previous epoch. The next epoch MUST have the same `alpha` with current receiver's `alpha`.

**Result:**

```txt
114997148547855332310731174935020155906209462858493962385407246111280193662921n
```

### .getReceiverEpoch()

```ts
await orandInstance.getReceiverEpoch(
  "0xf16f07cfd6e9ac06925fcf68dd0b450f4131989d"
);
```

Get current `epoch` of receiver, this value will be increased after a new epoch data submit to `OrandProviderV1`. This value MUST be equal to the epoch number of the submitting epoch.

**Result:**

```txt
1n
```

### .verifyECVRFProof()

```ts
await orandInstance.verifyECVRFProof(epochData);
```

> Allowed you to verify the correctness of any epoch, it will return `true` if the given epoch is valid otherwise it returns `false`.

Reviewer: ```suggestion
Allowed you to verify the correctness of any epoch, it will return `true` if the given epoch is valid, otherwise it will return `false`.
```
